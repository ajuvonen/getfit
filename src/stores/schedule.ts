import {defineStore} from 'pinia';
import {useStorage} from '@vueuse/core';
import {v4 as uuidv4} from 'uuid';
import {clone, lensProp, over} from 'ramda';
import {DateTime} from 'luxon';
import type {ScheduleSettings, Training, Week} from '@/types';
import {ACTIVITIES} from '@/constants';
import {roundNearestQuarter} from '@/utils';
import {computed, watch} from 'vue';

const getEmptySchedule = (): ScheduleSettings => ({
  name: '',
  startsOnSunday: false,
  startDate: null,
  actualWeekNumbering: false,
  availableActivities: ACTIVITIES.map(({value}) => value),
  defaultStartTime: {
    hours: 12,
    minutes: 0,
    seconds: 0,
  },
  defaultDuration: 1,
  unitOfTime: 'h',
});

export const useScheduleStore = defineStore('schedule', () => {
  // State refs
  const settings = useStorage('getfit-settings', getEmptySchedule(), localStorage, {
    mergeDefaults: true,
  });

  const weeks = useStorage('getfit-schedule', [] as Week[], localStorage, {mergeDefaults: true});

  // Computed getters
  const getTargetWeekAndTraining = computed(() => (weekId: string, trainingId?: string) => {
    const targetWeek = weeks.value.find(({id}) => id === weekId) as Week;
    const targetTraining = trainingId
      ? targetWeek.trainings.find(({id}) => id === trainingId)
      : undefined;
    return [targetWeek, targetTraining] as [Week, Training | undefined];
  });

  // Actions
  const addWeek = () => {
    weeks.value.push({
      id: uuidv4(),
      trainings: [],
    });
  };

  const deleteWeek = (weekId: string) => {
    weeks.value = weeks.value.filter(({id}) => id !== weekId);
  };

  const copyWeek = (weekId: string) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    const newWeekId = uuidv4();
    const cloneWeek: Week = {
      id: newWeekId,
      trainings: targetWeek.trainings.map((training) => ({
        ...training,
        id: uuidv4(),
        weekId: newWeekId,
      })),
    };
    weeks.value.push(cloneWeek);
  };

  const addOrEditTraining = (training: Training) => {
    const [targetWeek] = getTargetWeekAndTraining.value(training.weekId);
    const targetIndex = targetWeek.trainings.findIndex(({id}) => id === training.id);
    if (targetIndex >= 0) {
      targetWeek.trainings[targetIndex] = training;
    } else {
      targetWeek.trainings.push(training);
    }
  };

  const deleteTraining = (training: Training) => {
    const [targetWeek] = getTargetWeekAndTraining.value(training.weekId);
    targetWeek.trainings = targetWeek.trainings.filter(({id}) => id !== training.id);
  };

  const moveTraining = (training: Training, newWeekId: string, dayIndex: number) => {
    const [originalWeek, targetTraining] = getTargetWeekAndTraining.value(
      training.weekId,
      training.id,
    ) as [Week, Training];
    const [targetWeek] = getTargetWeekAndTraining.value(newWeekId);
    originalWeek.trainings = originalWeek.trainings.filter(({id}) => id !== targetTraining.id);
    targetTraining.weekId = newWeekId;
    targetTraining.dayIndex = dayIndex;
    targetWeek.trainings.push(targetTraining);
  };

  const reorderTrainings = (weekId: string, trainings: Training[]) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    targetWeek.trainings = trainings;
  };

  const copyTraining = (training: Training, weekId: string, dayIndex: number) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    const clonedTraining: Training = {
      ...clone(training),
      id: uuidv4(),
      weekId,
      dayIndex,
    };
    targetWeek.trainings.push(clonedTraining);
  };

  // Watchers
  watch(
    () => settings.value.unitOfTime,
    (newValue, oldValue) => {
      if (newValue && newValue !== oldValue) {
        const multiplier = newValue === 'h' ? 1 / 60 : 60;
        const precision = newValue === 'h' ? 2 : 0;
        settings.value.defaultDuration = roundNearestQuarter(
          settings.value.defaultDuration * multiplier,
          precision,
        );
        weeks.value = weeks.value.map((week) => ({
          ...week,
          trainings: week.trainings.map(
            over(lensProp('duration'), (duration) =>
              roundNearestQuarter(duration * multiplier, precision),
            ),
          ),
        }));
      }
    },
  );

  watch(
    () => settings.value.startsOnSunday,
    (newValue) => {
      const parsedValue = newValue || false;
      if (settings.value.startDate) {
        const date = DateTime.fromJSDate(settings.value.startDate);
        if (parsedValue) {
          settings.value.startDate = date.minus({days: 1}).toJSDate();
        } else {
          settings.value.startDate = date.plus({days: 1}).toJSDate();
        }
      }
    },
  );

  // Reset
  const $reset = () => {
    settings.value = getEmptySchedule();
    weeks.value = [];
  };

  return {
    settings,
    weeks,
    getTargetWeekAndTraining,
    addWeek,
    deleteWeek,
    copyWeek,
    addOrEditTraining,
    deleteTraining,
    moveTraining,
    reorderTrainings,
    copyTraining,
    $reset,
  };
});
