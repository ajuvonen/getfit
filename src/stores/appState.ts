import {defineStore} from 'pinia';
import {v4 as uuidv4} from 'uuid';
import type {Training} from '@/types';
import {clone} from 'ramda';
import {useScheduleStore} from './schedule';

export const useAppStateStore = defineStore('appState', {
  state: () => ({
    summaryShown: [] as string[],
    trainingDialogOpen: false,
    trainingData: {} as Training,
  }),
  getters: {
    isSummaryShown: (state) => (trainingId: string) => state.summaryShown.includes(trainingId),
  },
  actions: {
    openNewTrainingDialog(weekId: string, dayIndex: number) {
      this.trainingData = {
        id: uuidv4(),
        weekId,
        dayIndex,
        duration: 0,
        activity: '',
        description: '',
        title: '',
        intensity: 0,
        completionSummary: '',
        location: '',
      };
      this.trainingDialogOpen = true;
    },
    openEditTrainingDialog(training: Training) {
      const scheduleStore = useScheduleStore();
      const clonedTraining = clone(training);
      if (
        clonedTraining.activity &&
        !scheduleStore.schedule.availableActivities.includes(clonedTraining.activity)
      ) {
        clonedTraining.activity = '';
      }
      this.trainingData = clonedTraining;
      this.trainingDialogOpen = true;
    },
    closeEditTrainingDialog() {
      this.trainingDialogOpen = false;
    },
    toggleSummaryShown(trainingId: string) {
      if (this.isSummaryShown(trainingId)) {
        this.summaryShown = this.summaryShown.filter((id) => id !== trainingId);
      } else {
        this.summaryShown.push(trainingId);
      }
    },
  },
});
