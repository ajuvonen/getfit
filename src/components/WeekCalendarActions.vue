<script setup lang="ts">
import {useI18n} from 'vue-i18n';
import {useAppStateStore} from '@/stores/appState';
import {useScheduleStore} from '@/stores/schedule';
import useScreenSize from '@/hooks/screenSize';

defineProps<{
  weekId: string;
  weekNumber: number;
  dayIndex: number;
}>();

const scheduleStore = useScheduleStore();
const {deleteWeek, copyWeek} = scheduleStore;
const {openNewTrainingDialog} = useAppStateStore();
const {isSmallScreen} = useScreenSize();
const {t} = useI18n();
</script>
<template>
  <div class="week-calendar__actions d-flex mt-4" :class="{'flex-column': isSmallScreen}">
    <v-btn
      :data-test-id="`week-${weekNumber}-add-training-button`"
      prepend-icon="mdi-plus"
      variant="flat"
      @click="openNewTrainingDialog(weekId, dayIndex)"
      >{{ t('weekCalendar.addTraining') }}</v-btn
    >
    <v-btn
      :aria-label="t('weekCalendar.copyWeek', [weekNumber])"
      :data-test-id="`week-${weekNumber}-copy-button`"
      prepend-icon="mdi-content-copy"
      variant="flat"
      @click="copyWeek(weekId)"
      >{{ t('weekCalendar.copyWeek') }}</v-btn
    >
    <v-btn
      :aria-label="t('weekCalendar.deleteWeek', [weekNumber])"
      :data-test-id="`week-${weekNumber}-delete-button`"
      color="error"
      variant="outlined"
      prepend-icon="mdi-delete"
      @click="deleteWeek(weekId)"
      >{{ t('weekCalendar.deleteWeek') }}</v-btn
    >
  </div>
</template>
