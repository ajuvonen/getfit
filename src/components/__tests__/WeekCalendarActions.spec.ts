import {mount} from '@vue/test-utils';
import {v4 as uuid} from 'uuid';
import {describe, it, expect, beforeEach} from 'vitest';
import {useScheduleStore} from '@/stores/schedule';
import {useAppStateStore} from '@/stores/appState';
import WeekCalendarActions from '@/components/WeekCalendarActions.vue';

describe('WeekCalendarActions', () => {
  let scheduleStore: ReturnType<typeof useScheduleStore>;
  let appStateStore: ReturnType<typeof useAppStateStore>;

  beforeEach(() => {
    scheduleStore = useScheduleStore();
    appStateStore = useAppStateStore();
  });

  it('mounts', () => {
    const wrapper = mount(WeekCalendarActions, {
      props: {
        weekIndex: 0,
        dayIndex: 0,
        weekId: uuid(),
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('actions work', async () => {
    const weekId = uuid();
    const wrapper = mount(WeekCalendarActions, {
      props: {
        weekIndex: 0,
        dayIndex: 0,
        weekId: weekId,
      },
    });
    await wrapper.findByTestId('week-0-add-training-button').trigger('click');
    await wrapper.findByTestId('week-0-copy-button').trigger('click');
    await wrapper.findByTestId('week-0-delete-button').trigger('click');
    expect(appStateStore.openNewTrainingDialog).toHaveBeenCalledWith(
      weekId,
      0,
      expect.objectContaining([
        {
          title: 'Badminton',
          value: 'badminton',
        },
      ]),
    );
    expect(scheduleStore.copyWeek).toHaveBeenCalledWith(weekId);
    expect(scheduleStore.deleteWeek).toHaveBeenCalledWith(weekId);
  });
});
