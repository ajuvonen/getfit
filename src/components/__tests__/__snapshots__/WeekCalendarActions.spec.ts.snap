// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeekCalendarActions > mounts 1`] = `
"<div class="week-calendar__actions d-flex mt-4"><button type="button" class="v-btn v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-flat" data-test-id="week-1-add-training-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span><span class="v-btn__prepend"><i class="mdi-plus mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span><span class="v-btn__content" data-no-activator="">Add training</span>
    <!---->
    <!---->
  </button><button type="button" class="v-btn v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-flat" aria-label="Copy Week 1" data-test-id="week-1-copy-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span><span class="v-btn__prepend"><i class="mdi-content-copy mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span><span class="v-btn__content" data-no-activator="">Copy Week </span>
    <!---->
    <!---->
  </button><button type="button" class="v-btn v-theme--light text-error v-btn--density-default v-btn--size-default v-btn--variant-outlined" aria-label="Delete Week 1" data-test-id="week-1-delete-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span><span class="v-btn__prepend"><i class="mdi-delete mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span><span class="v-btn__content" data-no-activator="">Delete Week </span>
    <!---->
    <!---->
  </button></div>"
`;
